[RFC] Ensure build works without LuaJIT available

jamessan opened the pull request:
@teto ran into the below build failure, which this PR intends to fix (and cover
in CI to avoid regressions).

```
 -DCMAKE_INSTALL_PREFIX=/nix/store/szzyx7ay5j573s75wgiv421my2a8wvkb-neovim-local -DPREFER_LUA=ON -DLUA_PRG=/nix/store/8fm9n1w6janc5cvhn4dxnlkgdvflhxbx-lua-5.1.5-env/bin/lua -DLUA_PRG=/nix/store/8fm9n1w6janc5cvhn4dxnlkgdvflhxbx-lua-5.1.5-env/bin/lua -DCMAKE_BUILD_TYPE=Debug -DBUSTED_PRG=/nix/store/8fm9n1w6janc5cvhn4dxnlkgdvflhxbx-lua-5.1.5-env/bin/busted 
-- Replacing -O3 in CMAKE_C_FLAGS_RELEASE with -O2.
-- Could NOT find LuaJit (missing: LUAJIT_LIBRARY LUAJIT_INCLUDE_DIR) 
-- Found Iconv 
-- Checking Lua interpreter /nix/store/8fm9n1w6janc5cvhn4dxnlkgdvflhxbx-lua-5.1.5-env/bin/lua
-- Using the Lua interpreter /nix/store/8fm9n1w6janc5cvhn4dxnlkgdvflhxbx-lua-5.1.5-env/bin/lua.
-- MIN_LOG_LEVEL not specified
CMake Warning at CMakeLists.txt:573 (message):
  disabling unit tests: no Luajit FFI in
  /nix/store/8fm9n1w6janc5cvhn4dxnlkgdvflhxbx-lua-5.1.5-env/bin/lua
-- Configuring done
CMake Error at CMakeLists.txt:535 (file):
  Error evaluating generator expression:
    $<TARGET_FILE:nvim-test>
  No target "nvim-test"
CMake Error at CMakeLists.txt:535 (file):
  Error evaluating generator expression:
```

jamessan committed:
travis: Disable LuaJIT for functionaltest-lua build

Since we're already using Lua for the testing this allows us to ensure
our build still works properly without LuaJIT available.

jamessan committed:
third-party: luarocks: Use Lua if LuaJIT is disabled

jamessan committed:
third-party: lua: Set LUA_ROOT to ${DEPS_INSTALL_DIR}

jamessan committed:
third-party: Install luabitop if using Lua

jamessan committed:
cmake: Set TEST_LIBNVIM_PATH to empty string when not unit testing

jamessan commented:
CI now replicates the failure -- https://travis-ci.org/neovim/neovim/jobs/336223569.  My next commit should fix the issue.

jamessan committed:
travis: Don't run unit tests for functionaltest-lua build

justinmk commented on third-party/cmake/BuildLuarocks.cmake:
```diff
@@ -130,6 +130,18 @@ add_custom_target(inspect
+  add_custom_command(OUTPUT ${HOSTDEPS_LIB_DIR}/luarocks/rocks/luabitop
+    COMMAND ${LUAROCKS_BINARY}
+    ARGS build luabitop ${LUAROCKS_BUILDARGS}
+    DEPENDS inspect)
```
This is tricky, the `DEPENDS inspect` for penlight (below) could be parallel now.

I guess we need to check `if(USE_BUNDLED_LUA)` below, and set penlight `DEPENDS luabitop` in that case.

jamessan replied:
Or make `luabitop` the last item in the file.  That's probably easier.

jamessan committed:
test: man_spec: Fix use of nested [[ quoting

Lua (not LuaJIT) complains about the "^[[" strings inside the expect,
since it sees them as nested quotes.  Change the quoting to [=[ ]=] to
avoid the issue.

jamessan commented:
@ZyX-I Can you take a look at these failures when using Lua (not LuaJIT on the system) to run functional tests?

```
[  ERROR   ] test/unit/viml/expressions/parser_tests.lua @ 4850: api nvim_parse_expression works with strings
test/helpers.lua:502: bad argument #1 to 'format' (string expected, got table)
stack traceback:
	test/helpers.lua:502: in function <test/helpers.lua:485>
[  ERROR   ] test/unit/viml/expressions/parser_tests.lua @ 6972: api nvim_parse_expression works (KLEE tests)
test/helpers.lua:502: bad argument #1 to 'format' (string expected, got table)
stack traceback:
	test/helpers.lua:502: in function <test/helpers.lua:485>
```

ZyX-I commented:
Apparently this exact error is there because somebody issued `error({message=…})` and I did not really expect non-string error. Real error is

```
{
  message = [[test/functional/api/vim_spec.lua:943: Expected objects to be the same.
Passed in:
(table) {
 *[ast] = {
   *[1] = {
      [1] = 'Or:0:0:|'
     *[children] = {
        [1] = 'Missing:0:0:'
       *[2] = 'DoubleQuotedString(val="0:1:"\e"' } } }
  [err] = {
    [arg] = '|"\e"'
    [msg] = 'E15: Unexpected EOC character: %.*s' } }
Expected:
(table) {
 *[ast] = {
   *[1] = {
      [1] = 'Or:0:0:|'
     *[children] = {
        [1] = 'Missing:0:0:'
       *[2] = 'DoubleQuotedString(val="\27"):0:1:"\e"' } } }
  [err] = {
    [arg] = '|"\e"'
    [msg] = 'E15: Unexpected EOC character: %.*s' } }]]
}
```

ZyX-I committed:
functests: Improve error reporting in _check_parsing function

May be needed for unit tests as well though.

ZyX-I commented:
You can pull in error reporting fix from pr-7957 branch of my fork.

ZyX-I committed:
tests: Make format_string('%q', ...) output more stable

It appears to be different on lua and luajit.

ZyX-I commented:
@jamessan There is also commit with the actual tests fix.

jamessan commented:
> There is also commit with the actual tests fix.

There's only 1 new commit on your branch.  Did you forget to push the other one?

ZyX-I commented:
Looks like I did. Should be pushed now.

jamessan committed:
lint
